pkg/services/grpcserver/auth.go

```diff
@@ -0,0 +1,128 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	apikeygenprefix "github.com/grafana/grafana/pkg/components/apikeygenprefixed"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Authenticator can authenticate GRPC requests.
+type Authenticator struct {
+	logger   log.Logger
+	APIKey   apikey.Service
+	SQLStore sqlstore.Store
+}
+
+func NewAuthenticator(apiKey apikey.Service, sqlStore sqlstore.Store) *Authenticator {
+	return &Authenticator{
+		logger:   log.New("grpc-server-authenticator"),
+		APIKey:   apiKey,
+		SQLStore: sqlStore,
+	}
+}
+
+// Authenticate checks that a token exists and is valid. It stores the user
+// metadata in the returned context and removes the token from the context.
+func (a *Authenticator) Authenticate(ctx context.Context) (context.Context, error) {
+	return a.tokenAuth(ctx)
+}
+
+const tokenPrefix = "Bearer "
+
+func (a *Authenticator) tokenAuth(ctx context.Context) (context.Context, error) {
+	auth, err := extractAuthorization(ctx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !strings.HasPrefix(auth, tokenPrefix) {
+		return ctx, status.Error(codes.Unauthenticated, `missing "Bearer " prefix in "authorization" value`)
+	}
+
+	token := strings.TrimPrefix(auth, tokenPrefix)
+	if token == "" {
+		return ctx, status.Error(codes.Unauthenticated, "token required")
+	}
+
+	newCtx := purgeHeader(ctx, "authorization")
+
+	err = a.validateToken(ctx, token)
+	if err != nil {
+		logger.Warn("request with invalid token", "error", err, "token", token)
+		return ctx, status.Error(codes.Unauthenticated, "invalid token")
+	}
+	return newCtx, nil
+}
+
+func (a *Authenticator) validateToken(ctx context.Context, keyString string) error {
+	decoded, err := apikeygenprefix.Decode(keyString)
```

Comment:
what's the difference between this and `ctx.Value("authorization-token")`?

Reply:
this parses the API key string itself, so it's already available when this function is called